rust/src/dcerpc/dcerpc.rs

```diff
@@ -0,0 +1,2600 @@
+            }
+        };
+
+        if self.data_needed_for_dir != direction && buffer.len() != 0 {
+            SCLogDebug!("DCERPC FAILED check data_needed_for_dir");
+            return AppLayerResult::err();
+        }
+
+        // Set data_needed_for_dir in the same direction in case there is an issue with parsing
+        self.data_needed_for_dir = direction;
+
+        // Making it a hard requirement for a data stream to at least have the header in place
+        // else return -1 and the buffer stays empty
+        // Reason: header has the most important of data required for further proccessing
+        if self.bytes_consumed < 16 && input_len > 0 {
+            parsed = self.process_header(&buffer);
+            if parsed == -1 {
+                self.extend_buffer(buffer, direction);
+                return AppLayerResult::ok();
+                // ASK the case when header is incomplete and that's all the data, returning
+                // ok would probably be incorrect
+            }
+            self.bytes_consumed += parsed as u16;
+        }
+
+        let fraglen = match self.get_hdr_fraglen() {
+            Some(x) => x,
+            None => 0,
+        };
+
+        if (buffer.len() as u16) < fraglen {
+            SCLogDebug!("Possibly fragmented data, waiting for more..");
+            self.extend_buffer(buffer, direction);
+            return AppLayerResult::ok();
+        } else {
+            self.query_completed = true;
+        }
+        // TODO buffer extra data when input length is more than fraglen
+        parsed = self.bytes_consumed as i32;
+
+        match self.get_hdr_type() {
+            Some(x) => match x {
+                DCERPC_TYPE_BIND | DCERPC_TYPE_ALTER_CONTEXT => {
+                    retval = self.process_bind_pdu(&buffer[parsed as usize..]);
+                    if retval == -1 {
+                        return AppLayerResult::err();
+                    }
+                }
+                DCERPC_TYPE_BINDACK | DCERPC_TYPE_ALTER_CONTEXT_RESP => {
+                    retval = self.process_bindack_pdu(&buffer[parsed as usize..]);
+                    if retval == -1 {
+                        return AppLayerResult::err();
+                    }
+                }
+                DCERPC_TYPE_REQUEST => {
+                    retval = self.process_request_pdu(&buffer[parsed as usize..]);
+                    if retval == -1 {
+                        return AppLayerResult::err();
+                    }
+                }
+                DCERPC_TYPE_RESPONSE => {
+                    let mut response = DCERPCResponse::new();
+                    response.cmd = match self.get_hdr_type() {
+                        Some(x) => x,
+                        None => 2,
+                    };
+                    self.response = Some(response);
+                    retval = self.handle_common_stub(
+                        &buffer[parsed as usize..],
+                        0,
```

Comment:
Not sure I understand this zero here. The argument is `bytes_consumed`. Should that be set to `parsed`? I think `handle_common_stub` needs documentation of the purpose of the function args.

Reply:
I was deleting 16 (hardcoded header length) in `handle_common_stub` and was using this param for request's consumed data from nom parser. Suppose I could pass a 16 there and removed hardcoded header length from the function. Thanks.

Reply:
hmm running into problem doing that. Realized why I did it like this in the first place. I think documentation would be better. That bytes_consumed param is for bytes after the header.